### PR TITLE
feat: fetch decompression

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

File: lib/fetch/index.js [hunk 1]
```diff
@@ -1554,9 +1563,85 @@ function httpNetworkFetch (
...
 
+          const codings = headers.get('content-encoding')
+          if (codings === 'gzip' || codings === 'x-gzip') {
+            this.decoder = zlib.createGunzip()
+          } else if (codings === 'deflate' || codings === 'x-deflate') {
+            this.decoder = zlib.createInflate()
+          } else if (codings === 'br') {
+            this.decoder = zlib.createBrotliDecompress()
+          } else {
+            // TODO: What to do when coding is unsupported?
```
Comment (ronag) [inline lib/fetch/index.js:1574]: @mcollina @szmarczak What to do when coding is unsupported?

Comment (szmarczak) [reply lib/fetch/index.js:1574]: Got returns raw buffer. I guess `fetch` would do this as well. Let me check

Comment (ronag) [reply lib/fetch/index.js:1553]: Also it's a little weird that the end user has no way of determining whether the response has been decoded or not.

**Comment by ronag** [comment]
@szmarczak there is a node-fetch test that says:

> 'should make capitalised Content-Encoding lowercase'

But I can't find anything in the spec that says we should do that. wdyt?

**Comment by szmarczak** [comment]
> But I can't find anything in the spec that says we should do that. 

Me neither. It needs to be case-insensitive only when creating a list from it.

**Comment by szmarczak** [comment]
It should be left as it is. The only normalization going on is trimming.

**Comment by ronag** [comment]
I removed that test.

File: lib/fetch/index.js (hunk 1 above)
Comment (ronag) [reply lib/fetch/index.js:1558]: @szmarczak would you mind checking what chrome does?

Comment (szmarczak) [reply lib/fetch/index.js:1558]: Returns the value with no errors.

Comment (szmarczak) [reply lib/fetch/index.js:1558]: > Also it's a little weird that the end user has no way of determining whether the response has been decoded or not.

I agree. Can you ask someone from the `fetch` team about this?

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Ethan-Arrowood — APPROVED** [review]
this looks great! nice work

File: lib/fetch/index.js [hunk 2]
```diff
@@ -1562,78 +1556,155 @@ function httpNetworkFetch (
...
+              .map((x) => x.trim()) ?? []
+
+          const decoders = []
+
+          // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding
+          for (const coding of codings) {
+            if (/(x-)?gzip/.test(coding)) {
+              decoders.push(zlib.createGunzip())
+            } else if (/(x-)?deflate/.test(coding)) {
+              decoders.push(zlib.createInflate())
```
Comment (kanongil) [inline lib/fetch/index.js:1573]: Shouldn't this be `zlib.createDeflate()`?

File: lib/fetch/index.js [hunk 3]
```diff
@@ -1562,78 +1556,155 @@ function httpNetworkFetch (
...
+            } else if (coding === 'br') {
+              decoders.push(zlib.createBrotliDecompress())
+            } else {
+              // TODO: What to do when coding is invalid or unsupported?
```
Comment (kanongil) [inline lib/fetch/index.js:1577]: Spec says to [just passthrough](https://fetch.spec.whatwg.org/#handle-content-codings). Similar to the `identity` encoding.

This means any unknown non-`identity` encoding should invalidate this decoder chain.

**Review by kanongil — COMMENTED** [review]
Great, I looked at the fetch feature yesterday and wondered where the encodings where handled.

File: lib/fetch/index.js (hunk 3 above)
Comment (ronag) [reply lib/fetch/index.js:1577]: Would you mind opening a PR to fix?

File: lib/fetch/index.js (hunk 2 above)
Comment (ronag) [reply lib/fetch/index.js:1573]: inflate is actually correct

File: lib/fetch/index.js (hunk 3 above)
Comment (ronag) [reply lib/fetch/index.js:1577]: Fixed in https://github.com/nodejs/undici/commit/963ca033466cacbf6c72e299871f1e6aa4d1ced8

**Pull request merged**